articles/152_roslaunch_multi_machine_launch.md

```diff
@@ -0,0 +1,310 @@
+[INFO] [launch]: All log files can be found below /home/preed/.ros/log/2019-09-11-20-54-30-715383-regulus-2799
+[INFO] [launch]: Default logging verbosity is set to INFO
+[INFO] [launch]: Launch System ID is 50bda6fb-d451-4d53-8a2b-e8fcdce8170b
+$
+```
+
+#### `list`
+
+Since it is possible to launch a system of nodes that spans a network and detach
+from it, it is necessary to be able to query the network to find which systems
+are active.
+
+```bash
+$ ros2 launcher list -h
+usage: ros2 launcher list [-h] [-v] [--spin-time SPIN_TIME]
+
+List running launch systems
+
+
+optional arguments:
+  -h, --help            Show this help message and exit.
+  -v, --verbose         Provides more verbose output.
+  --spin-time SPIN_TIME
+                        Spin time in seconds to wait for discovery (only
+                        applies when not using an already running daemon)
+```
+
+Example output:
+
+```bash
+$ ros2 launcher list
+ab1e0138-bb22-4ec9-a590-cf377de42d0f
+50bda6fb-d451-4d53-8a2b-e8fcdce8170b
+5d186778-1f50-4828-9425-64cc2ed1342c
+$
+```
+
+```bash
+$ ros2 launcher list
+ab1e0138-bb22-4ec9-a590-cf377de42d0f: 5 nodes, 2 hosts
+50bda6fb-d451-4d53-8a2b-e8fcdce8170b: 2 nodes, 1 host
+5d186778-1f50-4828-9425-64cc2ed1342c: 16 nodes, 3 hosts
+$
+```
+
+#### `attach`
+
+Since it is possible to detach from a launched system, it is useful for
+scripting or diagnostic purposes to be able to re-attach to it.
+
+```bash
+$ ros2 launcher attach -h
+usage: ros2 launcher attach [-h] [-v] [--spin-time SPIN_TIME] [system_id]
+
+Blocks until all nodes running under the specified Launch System ID have exited
+
+positional arguments:
+  system_id          Launch System ID of the nodes to attach to; if less than
+                     a full UUID is specified, it will attach to the first
+                     Launch System it finds whose ID begins with that sub-string
+
+optional arguments:
+  -h, --help            Show this help message and exit.
+  -v, --verbose         Provides more verbose output.
+  --spin-time SPIN_TIME
+                        Spin time in seconds to wait for discovery (only
+                        applies when not using an already running daemon)
+```
+
+Example output:
+
+```bash
+$ ros2 launcher attach 50bda6fb-d451-4d53-8a2b-e8fcdce8170b
+Attached to Launch System 50bda6fb-d451-4d53-8a2b-e8fcdce8170b.
+(... in another terminal, run `ros2 launcher term 50bda6fb`...)
+All nodes in Launch System 50bda6fb-d451-4d53-8a2b-e8fcdce8170b have exited.
+$
+```
+
+Verbose mode:
+
+```bash
+$ ros2 launcher attach -v 50bda6fb
+Attached to Launch System 50bda6fb-d451-4d53-8a2b-e8fcdce8170b.
+Waiting for node /launch_ros
+Waiting for node /talker
+Waiting for node /listener
+(... in another terminal, run `ros2 launcher term 50bda6fb`...)
+Node /launch_ros has exited
+Node /talker has exited
+Node /listener has exited
+All nodes in Launch System 50bda6fb-d451-4d53-8a2b-e8fcdce8170b have exited.
+$
+```
+
+#### `term`
```

Comment:
These features all seem like extensions that can extend the current `ros2run` and `ros2launch` CLI.
They are also something I would like to see in ROS2 as well :)

Reply:
I agree it would possibly be a useful feature of `ros2 launch` (not sure about `ros2 run`, but maybe...), but again I don't see what it has to do with multi-machine launch. We could add `attach`/`detach`/`term` without adding multi-machine launching actions, and vice versa (unless I miss some core reason for it being included here).